language/vm/src/views.rs

```diff
@@ -531,63 +532,132 @@ impl<'a, T: ModuleAccess> SignatureTokenView<'a, T> {
         self.token.signature_token_kind()
     }
 
-    #[inline]
-    pub fn kind(&self) -> Kind {
+    /// Returns the kind of the signature token in the given context (module, function/struct).
+    /// The context is needed to determine the kinds of structs & type variables.
+    /// Also checks whether kind constraints are met for all struct instantiations recursively.
+    // TODO: refactor views so that we get the type formals from self.
+    pub fn kind(&self, type_formals: &[Kind]) -> Result<Kind, VMStaticViolation> {
+        use SignatureToken::*;
+
         match self.token {
-            // TODO: Type actuals are ignored, fix it (generics).
-            SignatureToken::Struct(sh_idx, type_arguments) => {
-                let is_nominal_resource = StructHandleView::new(self.module, self.module.struct_handle_at(*sh_idx))
-                    .is_nominal_resource();
-                // Nominal resources are always of kind `Kind::Resource`
-                if is_nominal_resource {
-                    return Kind::Resource
+            // These primitive types have kind unrestricted.
+            Bool | U64 | String | ByteArray | Address => Ok(Kind::Unrestricted),
+
+            // Reference types are unrestricted, but we still need to inspect its internals to
+            // ensure all instantiations satisfy kind constraints.
+            Reference(ty) | MutableReference(ty) => {
+                Self::new(self.module(), ty).kind(type_formals)?;
+                Ok(Kind::Unrestricted)
+            }
+
+            // To get the kind of a type parameter, we lookup its definition in the formals.
+            TypeParameter(idx) => match type_formals.get(*idx as usize) {
+                Some(kind) => Ok(*kind),
+                None => Err(VMStaticViolation::IndexOutOfBounds(
+                    IndexKind::TypeParameter,
+                    type_formals.len(),
+                    *idx as usize,
+                )),
+            },
+
+            Struct(idx, tys) => {
+                // Get the struct handle at idx. Note the index could be out of bounds.
+                let struct_handles = self.module().struct_handles();
+                let sh = struct_handles.get(idx.0 as usize).ok_or_else(|| {
+                    VMStaticViolation::IndexOutOfBounds(
+                        IndexKind::StructHandle,
+                        struct_handles.len(),
+                        idx.0 as usize,
+                    )
+                })?;
+
+                // Recursively verify all the type actuals and check if their kinds satisfy the
+                // constraints of the current struct.
+                let struct_type_formals = &sh.type_formals;
```

Comment:
I'm confused how this comment applies

Reply:
*Replying to your comment above since github does not allow me to reply to it directly.

Cool we've all reached an agreement. `fn kind` will return `Kind`. Constraint checking will go to `bytecode_verifier::signature.rs`, and index checking into bounds checker.

Reply:
Didn't clean it up while moving things around. Will remove. Good catch.